helm-v3/deisgn-proposal.md

```diff
@@ -0,0 +1,715 @@
+
+## Pushing Charts To A Repository
+
+Several registries have emerged with support for Helm charts including ChartMuseum and Quay (via Application Registries) along with extras built on top of other systems, such as object storage (e.g., see the [S3 Plugin](https://github.com/hypnoglow/helm-s3)). These registries are asking for and could benefit from a capability to push packages from the Helm client.
+
+To support registries there will be a `helm push` command that works in a similar manner to `helm fetch`. It will support different systems via the URI/URL scheme and provide a default implementation for HTTP(S). Other systems, such as S3, can implement an uploader via a plugin for schemes Helm does not provide an implementation for.
+
+> [name=Matt Farina] There very well may be additional work here around registry search and pluggable or otherwise extendable authenication mechanisms.
+> 
+> [name=Adnan Abdulhussein]
+> What would the default HTTP(S) implementation be?
+> [name=Matt Fisher]
+> Adnan, would you mind clarifying your comment a little?
+> 
+> [name=Adnan Abdulhussein]
+> Sorry, the proposal mentions there will be a default implemention for `helm push` to a regular HTTP(S) chart repository. It's unclear to me how this would actually work.
+> 
+> [name=Adnan Abdulhussein]
+> Sounds like this is something that will be fleshed out more in Josh's proposal, so I'll wait for that :)
+
+## Security Considerations
+
+This architecture for Helm is more secure than Helm 2 (and equally secure to Helm Classic). The following security goals are met by this proposal:
+
+- Helm now runs with the ID and permissions of the UserAccount or ServiceAccount executing the commands.
+- RBAC can now be applied to each user
+- On-the-wire security is as secure as the cluster is configured to be (e.g. if KUBECONFIG is set up for mTLS, that's what we use)
+- Auditability is now a feature of the cluster, as each chart is "created by" the user who runs the request
+- In-cluster CRDs can be secured with RBAC, too, providing access controls for releases.
+
+
+## Appendix A: A Helm Controller
+
+This section describes a second implementation of Helm. The goal of this implementation is to provide a controller/CRD façade for the core Helm features. _However, defining the specifics of this controller was deemed outside of the scope of this proposal, which focuses on the reference implementation of the Helm CLI_.
+
+A client-only Helm follows a push-based model. To facilitate a pull-based model, we can also create a Helm controller that provides a CRD facade, but uses the Helm library.
+
+It is suggested that this (a) be a core Helm project, but (b) is kept separately from the main Helm codebase.
+
+This controller accepts Chart and values configuration as a new `HelmRequest` CRD, and then provides the following operations:
+
+- Install
+- Upgrade
+- Delete
```

Comment:
-1. This is very imperative. In an operator like workflow, the user specifies what they want, and the operator keeps track of how to get it there. So they would say, originally request release 1.0.0 and then update the object to 1.0.1. the operator would then decide since it was 1.0.0 and now 1.0.1 that an upgrade is needed.

Install = upload new crd.
Upgrade = update the crud's fields.
Delete = delete the crd.

Reply:
I think you are mis-reading. The point is that the listed operations have to be handled by the controller. It must be able to install, upgrade, list, delete, etc. the releases. The front-end will be the typically one: a CRD.

We need to make it clear that _all_ of those features must be supported by the controller (not just an arbitrary subset). Otherwise, the functional disparity between the the controller and the client would require that one use both instead of just the controller.

Ultimately what we want is for operators to be able to decide to...

- Use just the CLI
- Use just the controller (with kubectl)
- Use both, and expose various restraints using RBAC

Consequently, for the second one to be true, we have to make it a requirement that the controller have support for at least those features.